src/util.rs

```diff
@@ -5,7 +5,280 @@
+        //     wrap around the address space, so does `ptr`.
+        // - `validate_cast_and_convert_metadata` promises that the object
+        //   described by `split_at` is validly-aligned for `U`.
+        // - By invariant on `self`, `A` is guaranteed to live for at least
+        //   `'a`.
+        // - `U: 'a` by trait bound.
+        Some((Ptr { ptr, _lifetime: PhantomData }, split_at))
+    }
+
+    /// Attempts to cast `self` into a `U`, failing if all of the bytes of
+    /// `self` cannot be treated as a `U`.
+    ///
+    /// In particular, this method fails if `self` is not validly-aligned for
+    /// `U` or if `self`'s size is not a valid size for `U`.
+    ///
+    /// # Safety
+    ///
+    /// On success, the caller may assume that the returned pointer references
+    /// the same byte range as `self`.
+    #[doc(hidden)]
+    #[inline(always)]
+    pub(crate) fn _try_cast_into_no_leftover<U: 'a + ?Sized + KnownLayout>(
+        &self,
+    ) -> Option<Ptr<'a, U>> {
+        // TODO(#67): Remove this allow. See NonNulSlicelExt for more details.
+        #[allow(unstable_name_collisions)]
+        match self._try_cast_into(_CastType::_Prefix) {
+            Some((slf, split_at)) if split_at == self._len() => Some(slf),
+            Some(_) | None => None,
+        }
+    }
+}
+
+impl<'a, T: 'a + ?Sized> From<&'a T> for Ptr<'a, T> {
+    #[inline(always)]
+    fn from(t: &'a T) -> Ptr<'a, T> {
+        // SAFETY: `t` points to a valid Rust allocation, `A`, by construction.
+        // Thus:
+        // - `ptr` is derived from `A`
+        // - Since we use `NonNull::from`, which preserves provenance, `ptr` has
+        //   the same provenance as `A`
+        // - Since `NonNull::from` creates a pointer which addresses the same
+        //   bytes as `t`, `ptr` addresses a byte range entirely contained in
+        //   (in this case, identical to) `A`
+        // - Since `t: &T`, it addresses no more than `isize::MAX` bytes. [1]
+        // - Since `t: &T`, it addresses a byte range which does not wrap around
+        //   the address space. [2]
+        // - Since it is constructed from a valid `&T`, `ptr` is validly-aligned
+        //   for `T`
+        // - Since `t: &'a T`, the allocation `A` is guaranteed to live for at
+        //   least `'a`
+        // - `T: 'a` by trait bound
+        //
+        // TODO(#429), TODO(https://github.com/rust-lang/rust/issues/116181):
+        // Once it's documented, reference the guarantee that `NonNull::from`
+        // preserves provenance.
+        //
+        // TODO(#429),
+        // TODO(https://github.com/rust-lang/unsafe-code-guidelines/issues/465):
+        // - [1] Where does the reference document that allocations fit in
+        //   `isize`?
+        // - [2] Where does the reference document that allocations don't wrap
+        //   around the address space?
+        Ptr { ptr: NonNull::from(t), _lifetime: PhantomData }
```

Comment:
It's regrettable that we don't yet have unsafe fields, or unsafe implicit instructors. What do you think about renaming the `ptr` field to `unsafe_ptr`, just to hammer the point home that this field has safety invariants at its use-sites?

Reply:
I'm worried that if we do that, then for consistency's sake we'll need to do it on almost every field in the crate. I feel like if you're working on this crate, you've gotta already be aware that most things have safety invariants (and certainly code reviewers know that), so the risk of missing an invariant is low. Another angle is that the sort of person who wouldn't know to check whether `ptr` has a safety invariant is the same sort of person who wouldn't get the hint from the name `unsafe_ptr`.

Reply:
Yeah fair enough; I've moved it into a submodule.